File: DictionaryProject/WordBuilder/RuleCollection.vb

Public Class RuleCollection
    Inherits Collections.ObjectModel.Collection(Of Rule)

    Private Shared _Random As New Random()

    Public Function GetRuleByName(ByVal name As String) As Rule
        Dim rules As New List(Of Rule)()
        Dim total As Double
        
        For Each r As Rule In Me
        	If r.Name = name Then
        		rules.Add(r)
        		total += r.Probability
        	End If
        Next

        If rules.Count = 0 Then
            Return Nothing
        End If

        Dim selected As Double = _Random.NextDouble() * total

        For Each r As Rule In rules
            selected -= r.Probability
            If selected <= 0 Then
                Return r
            End If
        Next

        Return rules.LastOrDefault()
    End Function
End Class

File: monowordbuilder/WordBuilderProject/RuleCollection.cs
using System;
using System.Collections.Generic;

public class RuleCollection : System.Collections.ObjectModel.Collection<Rule>
{
    
	private static Random _Random = new Random();
    
	public Rule GetRuleByName(string name)
	{
        List<Rule> rules = new List<Rule>();
        double total = 0;
        
        foreach (Rule r in this) {
            if (r.Name == name) {
                rules.Add(r);
                total += r.Probability;
            }
        }
        
     	if (rules.Count == 0) {
            return null;
        }
        
        double selected = _Random.NextDouble() * total;
        
        foreach (Rule r in rules) {
            selected -= r.Probability;
            if (selected <= 0) {
				return r;
            }
        }
        
        return rules.Count > 0 ? rules[rules.Count - 1] : null;
    }
}
